### PR TITLE
Improve char_lit's readability and speed

### PR DESCRIPTION
**Pull request by nnethercote**
This is my first contribution to rustc. Please let me know if I've done anything wrong. (I ran `make tidy` before making the pull request.)

### CODE REVIEW

File: src/libsyntax/parse/mod.rs [hunk 1]
```diff
@@ -318,7 +310,10 @@ pub fn char_lit(lit: &str) -> (char, isize) {
 
     let unicode_escape = || -> Option<(char, isize)> {
         if lit.as_bytes()[2] == b'{' {
-            let idx = lit.find('}').expect(msg2);
+            let idx = match lit.find('}') {
+                Some(idx2) => idx2,
+                None => panic!("lexer should have rejected a bad character escape {}", lit),
+            };
```
Comment (KiChjang) [inline src/libsyntax/parse/mod.rs]: I find this format more verbose and adds no new information.

File: src/libsyntax/parse/mod.rs [hunk 2]
```diff
@@ -329,12 +324,16 @@ pub fn char_lit(lit: &str) -> (char, isize) {
...
         'x' | 'X' => esc(4, lit),
         'u' => unicode_escape(),
         'U' => esc(10, lit),
         _ => None,
-    }.expect(msg2);
+    };
+    match res {
+        Some(res2) => res2,
+        None => panic!("lexer should have rejected a bad character escape {}", lit),
+    }
```
Comment (KiChjang) [inline src/libsyntax/parse/mod.rs]: Same here.

File: src/libsyntax/parse/mod.rs (hunk 1 above)
Comment (nnethercote) [inline src/libsyntax/parse/mod.rs]: That's true, but you're overlooking the speed improvement which is the motivation for the change.

Comment (jseyfried) [inline src/libsyntax/parse/mod.rs]: You could make this a bit less verbose while still keeping the speed improvement with `unwrap_or_else`:
```rust
let idx = lit.find('}').unwrap_or_else(|| {
    panic!("lexer should have rejected a bad character escape {}", lit)
});
```

**Comment by jseyfried** [comment]
@nnethercote Thanks for the PR!
r=me with `unwrap_or_else` instead of manually unwrapping with `match`.

**Comment by nnethercote** [comment]
@jseyfried: I'll try that. Should I modify the second patch, or make the change in an additional patch?

**Comment by jseyfried** [comment]
@nnethercote I would just amend the second commit.

**Comment by nnethercote** [comment]
I updated the second commit.  r? @jseyfried

**Pull request merged**